packages/shared/src/hooks/useEventListener.ts

```diff
@@ -0,0 +1,158 @@
+  [BroadcastChannel, BroadcastChannelEventMap],
+  [Document, DocumentEventMap],
+  [Element, ElementEventMap],
+  [EventSource, EventSourceEventMap],
+  [FileReader, FileReaderEventMap],
+  [GlobalEventHandlers, GlobalEventHandlersEventMap],
+  [HTMLBodyElement, HTMLBodyElementEventMap],
+  [HTMLElement, HTMLElementEventMap],
+  [HTMLMediaElement, HTMLMediaElementEventMap],
+  [IDBDatabase, IDBDatabaseEventMap],
+  [IDBOpenDBRequest, IDBOpenDBRequestEventMap],
+  [IDBRequest, IDBRequestEventMap],
+  [IDBTransaction, IDBTransactionEventMap],
+  [MediaDevices, MediaDevicesEventMap],
+  [MediaKeySession, MediaKeySessionEventMap],
+  [MediaQueryList, MediaQueryListEventMap],
+  [MediaSource, MediaSourceEventMap],
+  [MediaStream, MediaStreamEventMap],
+  [MediaStreamTrack, MediaStreamTrackEventMap],
+  [MessagePort, MessagePortEventMap],
+  [Notification, NotificationEventMap],
+  [OfflineAudioContext, OfflineAudioContextEventMap],
+  [PaymentRequest, PaymentRequestEventMap],
+  [Performance, PerformanceEventMap],
+  [PermissionStatus, PermissionStatusEventMap],
+  [RTCDTMFSender, RTCDTMFSenderEventMap],
+  [RTCDataChannel, RTCDataChannelEventMap],
+  [RTCDtlsTransport, RTCDtlsTransportEventMap],
+  [RTCPeerConnection, RTCPeerConnectionEventMap],
+  [SVGElement, SVGElementEventMap],
+  [SVGSVGElement, SVGSVGElementEventMap],
+  [ScreenOrientation, ScreenOrientationEventMap],
+  [ServiceWorker, ServiceWorkerEventMap],
+  [ServiceWorkerContainer, ServiceWorkerContainerEventMap],
+  [ServiceWorkerRegistration, ServiceWorkerRegistrationEventMap],
+  [SourceBuffer, SourceBufferEventMap],
+  [SourceBufferList, SourceBufferListEventMap],
+  [SpeechSynthesis, SpeechSynthesisEventMap],
+  [SpeechSynthesisUtterance, SpeechSynthesisUtteranceEventMap],
+  [TextTrack, TextTrackEventMap],
+  [TextTrackCue, TextTrackCueEventMap],
+  [TextTrackList, TextTrackListEventMap],
+  [WebSocket, WebSocketEventMap],
+  [Window, WindowEventMap],
+  [WindowEventHandlers, WindowEventHandlersEventMap],
+  [Worker, WorkerEventMap],
+  [XMLHttpRequest, XMLHttpRequestEventMap],
+  [XMLHttpRequestEventTarget, XMLHttpRequestEventTargetEventMap],
+];
+
+type MapDefinitionToEventMap<D, T> = {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  [K in keyof D]: D[K] extends [any, any]
+    ? T extends D[K][0]
+      ? D[K][1]
+      : never
+    : never;
+};
+type GetDOMEventMaps<T> = MapDefinitionToEventMap<DOMEventMapDefinitions, T>;
+
+type MapEventMapsToKeys<D> = {
+  [K in keyof D]: D[K] extends never ? never : keyof D[K];
+};
+type MapEventMapsToEvent<D, T extends PropertyKey> = {
+  [K in keyof D]: D[K] extends never
+    ? never
+    : T extends keyof D[K]
+    ? D[K][T]
+    : never;
+};
+
+interface GenericEventListener<T> {
+  (event: T): void;
+}
+
+export type TUseEventListenerOptions =
+  | boolean
+  | AddEventListenerOptions
+  | undefined;
+
+export interface MessageEventData {
+  eventKey?: string;
+}
+
+const useEventListener = <
```

Comment:
Whole hook and tests are mostly the same logic from [here](https://github.com/kickassCoderz/kickass-toolkit/blob/16d2b6264c7dd6b09b2412cc41ea9df0b7815f49/packages/react/src/hooks/useEventListener/useEventListener.ts), could not use the package since we are still on react@17.x so copied everything for now.